modules/phase_field/src/ics/SmoothCircleFromFileIC.C

```diff
@@ -0,0 +1,90 @@
+/****************************************************************/
+/* MOOSE - Multiphysics Object Oriented Simulation Environment  */
+/*                                                              */
+/*          All contents are licensed under LGPL V2.1           */
+/*             See LICENSE for full restrictions                */
+/****************************************************************/
+#include "SmoothCircleFromFileIC.h"
+
+template <>
+InputParameters
+validParams<SmoothCircleFromFileIC>()
+{
+  InputParameters params = validParams<SmoothCircleBaseIC>();
+  params.addClassDescription("Multiple smooth circles read from a text file");
+  params.addRequiredParam<FileName>("file_name", "File containing circle centers and radii");
+
+  return params;
+}
+
+SmoothCircleFromFileIC::SmoothCircleFromFileIC(const InputParameters & parameters)
+  : SmoothCircleBaseIC(parameters),
+    _data(0),
+    _file_name(getParam<FileName>("file_name")),
+    _txt_reader(_file_name, true, " ", &_communicator),
+    _n_circles(0)
+{
+  // Read names and vectors from file
+  _txt_reader.read();
+  _col_names = _txt_reader.getColumnNames();
+  _data = _txt_reader.getColumnData();
+
+  _n_circles = _data[0].size();
+
+  // Check that the file has all the correct information
+  bool col_x = false, col_y = false, col_z = false, col_r = false;
+  for (unsigned int i = 0; i < _col_names.size(); ++i)
+  {
+    if (_col_names[i] == "x")
+      col_x = true;
+    else if (_col_names[i] == "y")
+      col_y = true;
+    else if (_col_names[i] == "z")
+      col_z = true;
+    else if (_col_names[i] == "r")
+      col_r = true;
+
+    // Check that columns have uniform lengths
+    if (_data[i].size() != _n_circles)
+      mooseError("Columns in ", _file_name, " do not have uniform lengths.");
+  }
+
+  // Check that the required columns are present
+  if (col_x == false)
+    mooseError("No column in ", _file_name, " labeled 'x'.");
+  if (col_y == false)
+    mooseError("No column in ", _file_name, " labeled 'y'.");
+  if (col_z == false)
+    mooseError("No column in ", _file_name, " labeled 'z'.");
+  if (col_r == false)
+    mooseError("No column in ", _file_name, " labeled 'r'.");
```

Comment:
What I'd do in the constructor is create a map (well actually just a vector will do to map the columns in the file to the order you expect to work with them. For instance, If you want them in this order: r, x, y, z, but the user supplies them in this order: x, y, z, r. You'd use your "map" to give you back the corresponding index for the correct column. For this example your map would be like this `col_map = {3, 0, 1, 2}`. To make it really elegant you'd have an enum so you'd be accessing each column with the name of that column like this:

```
   enum class COLS { R, X, Y, Z};

   // Retrieve data from the R column
   _data[_col_map[COLS::R]]

   // Retrieve data from the X colun
  _data[_col_map[COLS::X]]
```